### ISSUE TITLE
speed to write files to s3 storage

### ISSUE THREAD

**Issue by tinaok** [pangeo-data/pangeo-eosc#39]
https://github.com/pangeo-data/pangeo-eosc/blob/main/notebooks/SwiftXarrayTest.ipynb
@sebastian-luna-valero @guillaumeeb 
I've tried to write the Zarr file following the example and I confirm it is very slow.

Is there any way to make it faster?

**Comment by sebastian-luna-valero** [comment]
I can check with CESNET after @guillaumeeb confirms issues.

**Comment by guillaumeeb** [comment]
@tinaok could you point towards the code cell that you fill is slow? I have to admit I didn't go into details about performances, but this seemed reasonable to me when I did those tests. 

If I remember correctly, this was between 0.5s and 2s per chunk, depending on chunk size.

**Comment by tinaok** [comment]
Thank you @guillaumeeb and @sebastian-luna-valero 
Here is the example I was making to teach everyone to write their Zarr file close to https://pangeo-clivar.vm.fedcloud.eu/jupyterhub/

[https://github.com/pangeo-data/clivar-2022/blob/main/tutorial/examples/notebooks/save_to_cesnet_bucket_zarr_s3fs.ipynb](https://github.com/pangeo-data/clivar-2022/blob/main/tutorial/examples/notebooks/save_to_cesnet_bucket_zarr_s3fs.ipynb)

**Comment by tinaok** [comment]
@guillaumeeb If you have better example than that, please place them at [https://github.com/pangeo-data/clivar-2022/blob/main/tutorial/examples/notebooks/](https://github.com/pangeo-data/clivar-2022/blob/main/tutorial/examples/notebooks/save_to_cesnet_bucket_zarr_s3fs.ipynb) and make pull request?

**Comment by guillaumeeb** [comment]
So, I've looked at the notebook and tried to reproduced. This is something I've experienced during my tests: when writing a Zarr file to S3 on CESNET, there is an non compressible time before the writes of chunks actually start. This time is about 30s.

From your example, we can see that it is not related to Dask. So this must be on Zarr or Xarray side. I'm not sure what is happening between the Python command and task graphs begin in Dask. Maybe it's the time it takes to create Metadata? But this seems really high. You can verify this by using a LocalCluster and watching the tasks stream.

Maybe it comes from CESNET, I don't know. Did you try the same code using another Cloud infrastructure?

What I can say, is that once the writes of chunks are started, the bandwidth seems reasonable, so bandwidth and latency of CESNET storage looks good when writing the chunks of data.

We can try simple writes from command line or with s3fs to see if it comes from CESNET, or from something internal to Xarray or Zarr, or maybe from a conjunction of both...

**Comment by guillaumeeb** [comment]
Hm, trying

```Python
ds.to_zarr("/tmp/myfile", mode='w', consolidated=True)
```

takes less than a second, so there is definitely something going on with Zarr and the object store.

**Comment by guillaumeeb** [comment]
I've also tried to sync the dataset from local storage to CESNET object store using `aws s3` cli:

```bash
 time aws s3 --endpoint https://object-store.cloud.muni.cz --profile pangeoswift sync /tmp/myfile s3://tmp/guillaumeeb/
upload: ../../tmp/myfile/.zattrs to s3://tmp/guillaumeeb/.zattrs   
upload: ../../tmp/myfile/lon/.zarray to s3://tmp/guillaumeeb/lon/.zarray
upload: ../../tmp/myfile/Tair/.zattrs to s3://tmp/guillaumeeb/Tair/.zattrs
upload: ../../tmp/myfile/Tair/.zarray to s3://tmp/guillaumeeb/Tair/.zarray
upload: ../../tmp/myfile/lat/0 to s3://tmp/guillaumeeb/lat/0       
upload: ../../tmp/myfile/lat/.zarray to s3://tmp/guillaumeeb/lat/.zarray
upload: ../../tmp/myfile/.zgroup to s3://tmp/guillaumeeb/.zgroup  
upload: ../../tmp/myfile/.zmetadata to s3://tmp/guillaumeeb/.zmetadata
upload: ../../tmp/myfile/lat/.zattrs to s3://tmp/guillaumeeb/lat/.zattrs
upload: ../../tmp/myfile/Tair/0.0.0 to s3://tmp/guillaumeeb/Tair/0.0.0
upload: ../../tmp/myfile/lon/0 to s3://tmp/guillaumeeb/lon/0  
upload: ../../tmp/myfile/time/0 to s3://tmp/guillaumeeb/time/0
upload: ../../tmp/myfile/time/.zattrs to s3://tmp/guillaumeeb/time/.zattrs
upload: ../../tmp/myfile/time/.zarray to s3://tmp/guillaumeeb/time/.zarray
upload: ../../tmp/myfile/lon/.zattrs to s3://tmp/guillaumeeb/lon/.zattrs

real    0m1.989s
user    0m0.733s
sys     0m0.184s
```

Less than 2 seconds, so the problem does not come from CESNET per se. We're in the last situation, conjunction between using Zarr/Xarray and the object store. We should try the notebook on another system with S3 storage if possible. Else, maybe ask some help on Pangeo community side.

**Comment by tinaok** [comment]
Thank you very much @guillaumeeb !!  I would like to start letting students use the object storage for writing their Zarr file.  
I'll may be then warn student if it is small netcdf file, they should write it in netcdf format? And use Zarr only for big Zarr tmp file writing until we have definitive solution.  

Do you have a way for avoiding one students deleting other student's data?

**Comment by tinaok** [comment]
> Did you try the same code using another Cloud infrastructure?

I have access to https://us-central1-b.gcp.pangeo.io/ but I do not have access to it's s3 storage so I can not bench.  May be @rabernat can help us?

**Comment by tinaok** [comment]
Thank you @rabernat but I do not understand yet what I should change in this notebook to use Zarr.storage.FSStore ...

Also, we have one more question that we want to ask how US Pangeo is doing https://github.com/pangeo-data/pangeo-eosc/issues/17#issuecomment-1277768559

**Comment by rabernat** [comment]
To profile this, here is what I am doing. From my notebook, I run

```
%prun -D zarr.prof ds.to_zarr(f'{SCRATCH_BUCKET}/air_temperature.zarr')
```

Then I download the`zarr.prof` file and visualize it locally with `python -m snakeviz zarr.prof`. This produces a view like this:
![Screen Shot 2022-10-13 at 11 29 16 AM](https://user-images.githubusercontent.com/1197350/195639875-a3f871c9-c89d-4f44-8665-f1c359f6e8a8.png)

[zarr.prof.gz](https://github.com/pangeo-data/pangeo-eosc/files/9778285/zarr.prof.gz) (gzipped because github doesn't allow just `.prof` files.)


What I have learned from this is that an awful lot of time is being spent on calling `__contains__` on the store. In other words, Zarr is checking, maybe too much, what data are already stored. If the list operation on your object store is slow, this will take even more time.

I would be very interested to see the same profiling for your object store. Feel free to run the same command and post your results here.

This seems like a good target for optimization within Zarr. I'm pinging @martindurant, and expert on all of these things, to see if he can see any low-hanging fruit.

**Comment by guillaumeeb** [comment]
I've tried switching from `S3Map` to `FSStore` with no luck, the time taken is about the same, between 40 and 45s!

So I profiled the call, an here is what I get:
![image](https://user-images.githubusercontent.com/17138587/195660539-b4f1ea3a-11a0-4527-abd6-bd4e4188e2e1.png)

The profiling file.
[zarr.prof.gz](https://github.com/pangeo-data/pangeo-eosc/files/9779011/zarr.prof.gz)

I'm not sure how to analyze this, it seems that every steps take longer time. I guess this is probably due to a higher latency of some commands?

**Comment by tinaok** [comment]
@rabernat @guillaumeeb Thank you.

I'm still stuck with saving Zarr file FSStore 

I tried 
```
import zarr
access_key = !aws configure get aws_access_key_id
access_key = access_key[0]
secret_key = !aws configure get aws_secret_access_key
secret_key = secret_key[0]
your_name='tinaok'
s3_prefix =  "tmp/"+your_name
zarr_file_name= "newzarr"
uri = f"s3://{s3_prefix}/{zarr_file_name}"
storage_kwargs={
         "key":access_key,
         "secret":secret_key}
store = zarr.storage.FSStore(uri, **storage_kwargs)
```
@guillaumeeb How did you specify `'endpoint_url': 'https://object-store.cloud.muni.cz' ` ?

**Comment by rabernat** [comment]
Here's a useful breakdown of your profile.

A huge amount of time is being spent on `__contains__` calls. I am quite certain we can optimize this. Next step is to start looking at the zarr source code more carefully.

ncalls | tottime | percall | cumtime | percall | filename:lineno(function)
-- | -- | -- | -- | -- | --
1 | 7.953e-05 | 7.953e-05 | 46.81 | 46.81 | ~:0(<built-in method builtins.exec>)
1 | 4.942e-05 | 4.942e-05 | 46.81 | 46.81 | api.py:1511(to_zarr)
1 | 2.793e-05 | 2.793e-05 | 46.81 | 46.81 | <string>:1(<module>)
1 | 1.597e-05 | 1.597e-05 | 46.81 | 46.81 | dataset.py:1935(to_zarr)
105 | 0.002647 | 2.521e-05 | 46.71 | 0.4448 | asyn.py:60(sync)
105 | 0.001294 | 1.232e-05 | 46.71 | 0.4449 | asyn.py:108(wrapper)
105 | 0.001647 | 1.569e-05 | 46.69 | 0.4446 | threading.py:280(wait)
106 | 0.001035 | 9.767e-06 | 46.69 | 0.4404 | threading.py:563(wait)
421 | 46.68 | 0.1109 | 46.68 | 0.1109 | ~:0(<method 'acquire' of '_thread.lock' objects>)
1 | 8.417e-05 | 8.417e-05 | 43.37 | 43.37 | zarr.py:502(store)
1 | 8.858e-06 | 8.858e-06 | 43.37 | 43.37 | api.py:1235(dump_to_store)
1 | 0.000419 | 0.000419 | 31.93 | 31.93 | zarr.py:580(set_variables)
33 | 0.000313 | 9.483e-06 | 24.07 | 0.7295 | storage.py:1411(__contains__)
33 | 0.0002367 | 7.173e-06 | 24.07 | 0.7294 | mapping.py:171(__contains__)
4 | 0.0001056 | 2.64e-05 | 16.42 | 4.104 | creation.py:19(create)
4 | 7.26e-05 | 1.815e-05 | 16.42 | 4.104 | hierarchy.py:1034(_create_nosync)
4 | 6.31e-05 | 1.578e-05 | 16.42 | 4.104 | hierarchy.py:786(_write_op)
4 | 3.062e-05 | 7.655e-06 | 16.42 | 4.104 | hierarchy.py:1029(create)
4 | 7.684e-05 | 1.921e-05 | 14.97 | 3.743 | storage.py:283(init_array)
19 | 0.0001746 | 9.191e-06 | 12.4 | 0.6524 | storage.py:101(contains_array)
8 | 0.0001491 | 1.864e-05 | 11.71 | 1.463 | hierarchy.py:367(__contains__)
18 | 0.0002274 | 1.263e-05 | 11.68 | 0.6487 | storage.py:109(contains_group)
28 | 0.0003081 | 1.1e-05 | 10.26 | 0.3663 | storage.py:1366(__getitem__)
28 | 0.0002857 | 1.02e-05 | 10.26 | 0.3663 | mapping.py:133(__getitem__)
4 | 0.0002816 | 7.039e-05 | 9.202 | 2.3 | storage.py:427(_init_array_metadata)
11 | 0.0003803 | 3.457e-05 | 8.249 | 0.7499 | storage.py:1379(__setitem__)
4 | 7.747e-05 | 1.937e-05 | 6.561 | 1.64 | common.py:149(add)
4 | 0.0002106 | 5.264e-05 | 6.56 | 1.64 | core.py:1750(_set_selection)
4 | 4.959e-05 | 1.24e-05 | 6.56 | 1.64 | core.py:1260(__setitem__)

**Comment by martindurant** [comment]
I think the problem is, that a directory listing is cached by the s3fs instance, and makes further lookups much faster. But exists() (used by `__contains__`) in the absence of a cache will not call list but instead call head on the one item. This is much faster for the one item, but slower in the long run for many items. Also, while the `_exists()` method is async, and so many could be called concurrently, there is no simple API for checking the existence of many files in bulk - this does it serially. zarr isn't using the listdir method of the store, I'm not sure when that would get called.

A workaround would be to call `.find()` on the filesystem object before processing, which populates the directory cache all the way down, in a single call.
```
store.map.fs.find(rootdir);
```

**Comment by guillaumeeb** [comment]
> A workaround would be to call .find() on the filesystem object before processing, which populates the directory cache all the way down, in a single call.

Do you mean before calling `ds.to_zarr`?

The thing here is that the store is cleared, either by doing it manually before writing the zarr dataset, or either by Zarr I guess?

Anyway, clearing the store manually does give speedup. And on a store cleared, `store.map.fs.find(uri)` just returns an empty list.

Sorry @martindurant, I'm not sure if I've correctly understood your suggestion.

In any case, as we are starting from a free bucket/uri, Zarr should not have to call `__contains__` so many times, right?

**Comment by martindurant** [comment]
> the store is cleared

That's a good point! In this case, zarr should know that there are no keys available at all. 

Actually, by writing to the directory, you invalidate the cache anyway, so my suggestion would only be true for reading, not writing. 

I wonder whether s3fs should attempt to update *update* the relevant part of the directory cache when completing a write, rather than invalidating it. I have checked, and AWS doesn't quite give you the same information on finalizing a file as when querying it afterwards.

**Comment by tinaok** [comment]
@martindurant @rabernat 
would you like to have a connection to the cluster?
If yes plz 
~~follow https://pangeo-data.github.io/clivar-2022/before/setup.html#running-the-tutorial-on-the-european-open-science-cloud-eosc~~
TLDR
click
[https://aai-dev.egi.eu/registry/co_petitions/start/coef:290](https://aai-dev.egi.eu/registry/co_petitions/start/coef:290)
and
[https://aai.egi.eu/registry/co_petitions/start/coef:386](https://aai.egi.eu/registry/co_petitions/start/coef:111)
“statement for purpose” you can just say "hello"

**Comment by rabernat** [comment]
No thanks! I am not going to log in your cluster. 🙃  But I am happy to interact here on GitHub around optimizing Zarr and fsspec to improve your use case.

**Comment by tinaok** [comment]
Thank you @rabernat !

**Comment by tinaok** [comment]
Hi @martindurant and @guillaumeeb 

I tried [with FSStore and  find (51 s)](https://github.com/pangeo-data/clivar-2022/blob/fix-awth/tutorial/examples/notebooks/save_to_cesnet_bucket_zarr.ipynb) and [with FSStore and with out find (51.3 s)](https://github.com/pangeo-data/clivar-2022/blob/fix-awth/tutorial/examples/notebooks/save_to_cesnet_bucket_zarr-without_find.ipynb)
Compare to using [s3fs (52.4 s)](https://github.com/pangeo-data/clivar-2022/blob/main/tutorial/examples/notebooks/save_to_cesnet_bucket_zarr_s3fs.ipynb)
So I could not observe the speed up.   May be I am doing something wrong? 
Thank you

**Comment by guillaumeeb** [comment]
@tinaok this is also what I observed above, no speed up with these changes. So we've profiled the calls to try to understand why it takes such time on this object store.

**Comment by martindurant** [comment]
I think the question is, when zarr wants to write a chunk, why does it first need to check if that key already exists? We mean to overwrite anyway, so it shouldn't need to know. If we were instead updating part of a chunk, we would need to first read the values from the part of the chunk that is not being changed - but not in this case.

**Comment by rabernat** [comment]
> when zarr wants to write a chunk, why does it first need to check if that key already exists?

Now we are asking the right questions. 👌 

@tinaok & @guillaumeeb - this is how Pangeo has made progress over the past years: carefully identifying performance bottlenecks and working systematically to fix them **upstream** in the open source libraries. If you have the time and interest, I really encourage you to dig into the zarr source code and see if you can optimize this a bit. It's very satisfying! One behalf of the zarr developers, I can say we will gladly welcome a PR on this topic.

I also think there is something weird going on with your object store. Those list operations are going very slow compared to GCS or AWS S3. So perhaps there is a chance to optimize on that side as well.

If we can optimize both, I believe we can achieve a 10 - 100x speedup of this workflow.

**Comment by tinaok** [comment]
Thank you @rabernat @martindurant 
> I also think there is something weird going on with your object store. Those list operations are going very slow compared to GCS or AWS S3. So perhaps there is a chance to optimize on that side as well.

CESNET is based on openstack, would it be other native protocols I should use to make it faster may be? (or do we see slow down due to that in the trace file?)

**Comment by rabernat** [comment]
I went down a bit of a rabbit hole with this and learned some interesting things.

To understand what was happening better, I created a metaclass that would help me log everything the `FSStore` object was doing:

<details>

```python
import logging
import sys
import os
from functools import wraps, partial
from types import FunctionType, MethodType, MethodWrapperType
from time import perf_counter

from zarr.storage import FSStore

def format_args(args):
    if not isinstance(args, tuple):
        args = (args,)
    return ', '.join((
            repr(a)[:32]
            if not isinstance(a, bytes)
            else f'{len(a)} BYTES'
            for a in args
        ))

def wrapper(method):
    @wraps(method)
    def _impl(self, *method_args, **method_kwargs):
        tic = perf_counter()
        method_output = method(self, *method_args, **method_kwargs)
        toc = perf_counter()
        elapsed = 1000*(toc - tic)
        arg_reprs = format_args(method_args)
        output_reprs = (' -> ' + format_args(method_output)) if method_output else ''
        self.logger.info(f"{method.__name__}({arg_reprs}){output_reprs} - {elapsed:2.5f} ms") # ({method_args}, {method_kwargs})")
        return method_output
    return _impl

            
class LoggingMetaclass(type):
        
    def __new__(cls, clsname, bases, attrs, methods_to_log=()):
        
        instance = super().__new__(cls, clsname, bases, attrs)
        #instance = type(clsname, bases, attrs)
        
        methods_to_overwrite = (
            method_name
            for method_name in dir(instance)
            if callable(getattr(instance, method_name, None))
            and method_name in methods_to_log
        )
   
        for method_name in methods_to_overwrite:
            setattr(instance, method_name, wrapper(getattr(instance, method_name)))
  
        logging.basicConfig(level=logging.INFO, stream=sys.stdout)
        instance.logger = logging.getLogger(clsname)
        instance.logger.info("initialized")

        return instance

class Foo:
    
    def __init__(self, a):
        self.a = a

    def get_a(self):
        return self.a
    
    def __getitem__(self, key):
        if key=="a":
            return self.a
        else:
            raise KeyError
    
class LoggingFoo(Foo, metaclass=LoggingMetaclass, methods_to_log=['get_a', '__getitem__']):
    pass

lf = LoggingFoo(a=1)
lf.get_a()
lf['a']

store_methods = [
    'getitems', '__getitem__', 'setitems', '__setitem__',
    'delitems', '__delitem__', '__contains__', 'keys'
]


class FinalMeta(type(FSStore), LoggingMetaclass):
    pass

class LoggedFSStore(FSStore, metaclass=FinalMeta, methods_to_log=store_methods):
    pass
```

</details>

With this I could watch the action as follows

```python
import xarray as xr
import gcsfs

fs = gcsfs.GCSFileSystem()
SCRATCH_BUCKET = os.environ['SCRATCH_BUCKET'] 
print(SCRATCH_BUCKET)
try:
    # make sure scratch dir is empty
    fs.rm(SCRATCH_BUCKET, recursive=True)
except FileNotFoundError:
    pass

ds = xr.tutorial.open_dataset("air_temperature.nc").load()
encoding = {var: {"chunks": ds[var].shape} for var in ds.variables}
store = LoggedFSStore(f'{SCRATCH_BUCKET}/air_temperature.zarr')
ds.to_zarr(store, encoding=encoding, mode='w')
```

which gave me the following logs

```
INFO:LoggedFSStore:__setitem__('.zgroup', 24 BYTES) - 161.02744 ms
INFO:LoggedFSStore:__contains__('.zarray') - 134.78823 ms
INFO:LoggedFSStore:__getitem__('.zgroup') -> 24 BYTES - 43.74526 ms
INFO:LoggedFSStore:__contains__('lat/.zarray') - 76.33124 ms
INFO:LoggedFSStore:__contains__('lat/.zgroup') - 73.61064 ms
INFO:LoggedFSStore:__contains__('air/.zarray') - 78.77026 ms
INFO:LoggedFSStore:__contains__('air/.zgroup') - 78.51281 ms
INFO:LoggedFSStore:__contains__('lon/.zarray') - 79.81741 ms
INFO:LoggedFSStore:__contains__('lon/.zgroup') - 87.17914 ms
INFO:LoggedFSStore:__contains__('time/.zarray') - 78.06299 ms
INFO:LoggedFSStore:__contains__('time/.zgroup') - 78.81959 ms
INFO:LoggedFSStore:__setitem__('.zattrs', 307 BYTES) - 141.40278 ms
INFO:LoggedFSStore:__contains__('lat/.zarray') - 74.32906 ms
INFO:LoggedFSStore:__contains__('lat/.zgroup') - 72.73923 ms
INFO:LoggedFSStore:__contains__('.zarray') - 83.58116 ms
INFO:LoggedFSStore:__contains__('.zgroup') -> True - 59.21848 ms
INFO:LoggedFSStore:__contains__('lat/.zarray') - 77.10479 ms
INFO:LoggedFSStore:__contains__('lat/.zgroup') - 80.31471 ms
INFO:LoggedFSStore:__setitem__('lat/.zarray', 315 BYTES) - 149.82518 ms
INFO:LoggedFSStore:__getitem__('lat/.zarray') -> 315 BYTES - 44.97753 ms
INFO:LoggedFSStore:__setitem__('lat/.zattrs', 159 BYTES) - 143.53136 ms
INFO:LoggedFSStore:setitems({'lat/0': b'\x02\x013\x04d\x00\x) - 65.50918 ms
INFO:LoggedFSStore:__contains__('lon/.zarray') - 78.73831 ms
INFO:LoggedFSStore:__contains__('lon/.zgroup') - 69.94445 ms
INFO:LoggedFSStore:__contains__('.zarray') - 75.09015 ms
INFO:LoggedFSStore:__contains__('.zgroup') -> True - 82.24049 ms
INFO:LoggedFSStore:__contains__('lon/.zarray') - 78.93038 ms
INFO:LoggedFSStore:__contains__('lon/.zgroup') - 78.95320 ms
INFO:LoggedFSStore:__setitem__('lon/.zarray', 315 BYTES) - 141.58281 ms
INFO:LoggedFSStore:__getitem__('lon/.zarray') -> 315 BYTES - 30.87437 ms
INFO:LoggedFSStore:__setitem__('lon/.zattrs', 160 BYTES) - 148.20924 ms
INFO:LoggedFSStore:setitems({'lon/0': b'\x02\x011\x04\xd4\x0) - 69.60758 ms
INFO:LoggedFSStore:__contains__('time/.zarray') - 72.39811 ms
INFO:LoggedFSStore:__contains__('time/.zgroup') - 81.09918 ms
INFO:LoggedFSStore:__contains__('.zarray') - 82.42646 ms
INFO:LoggedFSStore:__contains__('.zgroup') -> True - 47.30075 ms
INFO:LoggedFSStore:__contains__('time/.zarray') - 75.26746 ms
INFO:LoggedFSStore:__contains__('time/.zgroup') - 72.75635 ms
INFO:LoggedFSStore:__setitem__('time/.zarray', 318 BYTES) - 141.51360 ms
INFO:LoggedFSStore:__getitem__('time/.zarray') -> 318 BYTES - 38.26456 ms
INFO:LoggedFSStore:__setitem__('time/.zattrs', 192 BYTES) - 140.15968 ms
INFO:LoggedFSStore:setitems({'time/0': b'\x02\x01!\x08@[\x00) - 68.22306 ms
INFO:LoggedFSStore:__contains__('air/.zarray') - 70.68645 ms
INFO:LoggedFSStore:__contains__('air/.zgroup') - 60.35751 ms
INFO:LoggedFSStore:__contains__('.zarray') - 78.56590 ms
INFO:LoggedFSStore:__contains__('.zgroup') -> True - 58.17347 ms
INFO:LoggedFSStore:__contains__('air/.zarray') - 81.18960 ms
INFO:LoggedFSStore:__contains__('air/.zgroup') - 70.30224 ms
INFO:LoggedFSStore:__setitem__('air/.zarray', 367 BYTES) - 163.85888 ms
INFO:LoggedFSStore:__getitem__('air/.zarray') -> 367 BYTES - 49.24272 ms
INFO:LoggedFSStore:__setitem__('air/.zattrs', 471 BYTES) - 150.49870 ms
INFO:LoggedFSStore:setitems({'air/0.0.0': b'\x02\x01!\x04 %\) - 242.37073 ms
INFO:LoggedFSStore:keys() -> <generator object FSMap.__iter__ - 24.50295 ms
INFO:LoggedFSStore:__getitem__('.zattrs') -> 307 BYTES - 35.04010 ms
INFO:LoggedFSStore:__getitem__('.zgroup') -> 24 BYTES - 27.86014 ms
INFO:LoggedFSStore:__getitem__('air/.zarray') -> 367 BYTES - 34.22666 ms
INFO:LoggedFSStore:__getitem__('air/.zattrs') -> 471 BYTES - 31.78257 ms
INFO:LoggedFSStore:__getitem__('lat/.zarray') -> 315 BYTES - 25.75439 ms
INFO:LoggedFSStore:__getitem__('lat/.zattrs') -> 159 BYTES - 45.25747 ms
INFO:LoggedFSStore:__getitem__('lon/.zarray') -> 315 BYTES - 38.31415 ms
INFO:LoggedFSStore:__getitem__('lon/.zattrs') -> 160 BYTES - 33.82616 ms
INFO:LoggedFSStore:__getitem__('time/.zarray') -> 318 BYTES - 27.00831 ms
INFO:LoggedFSStore:__getitem__('time/.zattrs') -> 192 BYTES - 35.61871 ms
INFO:LoggedFSStore:__setitem__('.zmetadata', 3989 BYTES) - 65.72519 ms
INFO:LoggedFSStore:__getitem__('.zmetadata') -> 3989 BYTES - 62.98140 ms
```

It is RIDICULOUS how much unnecessary listing of the same objects over and over is being done.

I am confident that we can pretty easily get a 10x improvement here by refactoring Zarr.

**Comment by rabernat** [comment]
Out of curiosity, I tried implementing a very naive implementation of storing the data which bypasses most of xarray. This allows us to see whether xarray itself is responsible for these inefficient calls

```python
import zarr
from xarray.backends.zarr import encode_zarr_attr_value

def naive_store(ds, store):
    group = zarr.open_group(store, mode="w")
    group.attrs.update(ds.attrs)
    for v in ds.variables:
        a = group.create(v, shape=ds[v].shape, chunks=ds[v].shape, dtype=ds[v].dtype)
        a[:] = ds[v].values
        attrs = {k: encode_zarr_attr_value(val) for k, val in ds[v].attrs.items()}
        attrs['_ARRAY_DIMENSIONS'] = ds[v].dims
        a.attrs.update(attrs)
        
%time naive_store(ds, LoggedFSStore(f'{SCRATCH_BUCKET}/air_temperature_naive.zarr'))
```

```
INFO:LoggedFSStore:__setitem__('.zgroup', 24 BYTES) - 126.42858 ms
INFO:LoggedFSStore:__contains__('.zarray') - 65.40666 ms
INFO:LoggedFSStore:__getitem__('.zgroup') -> 24 BYTES - 32.38567 ms
INFO:LoggedFSStore:__setitem__('.zattrs', 307 BYTES) - 190.39598 ms
INFO:LoggedFSStore:__contains__('.zarray') - 67.66061 ms
INFO:LoggedFSStore:__contains__('.zgroup') -> True - 40.53481 ms
INFO:LoggedFSStore:__contains__('lat/.zarray') - 86.53445 ms
INFO:LoggedFSStore:__contains__('lat/.zgroup') - 70.08551 ms
INFO:LoggedFSStore:__setitem__('lat/.zarray', 313 BYTES) - 136.86775 ms
INFO:LoggedFSStore:__getitem__('lat/.zarray') -> 313 BYTES - 33.09395 ms
INFO:LoggedFSStore:setitems({'lat/0': b'\x02\x013\x04d\x00\x) - 98.51932 ms
INFO:LoggedFSStore:__setitem__('lat/.zattrs', 159 BYTES) - 105.16701 ms
INFO:LoggedFSStore:__contains__('.zarray') - 73.29439 ms
INFO:LoggedFSStore:__contains__('.zgroup') -> True - 41.47805 ms
INFO:LoggedFSStore:__contains__('air/.zarray') - 70.57624 ms
INFO:LoggedFSStore:__contains__('air/.zgroup') - 76.24795 ms
INFO:LoggedFSStore:__setitem__('air/.zarray', 365 BYTES) - 161.10498 ms
INFO:LoggedFSStore:__getitem__('air/.zarray') -> 365 BYTES - 34.90320 ms
INFO:LoggedFSStore:setitems({'air/0.0.0': b'\x02\x01!\x04 %\) - 214.99064 ms
INFO:LoggedFSStore:__setitem__('air/.zattrs', 471 BYTES) - 116.65585 ms
INFO:LoggedFSStore:__contains__('.zarray') - 64.61574 ms
INFO:LoggedFSStore:__contains__('.zgroup') -> True - 43.48885 ms
INFO:LoggedFSStore:__contains__('lon/.zarray') - 78.02696 ms
INFO:LoggedFSStore:__contains__('lon/.zgroup') - 99.15074 ms
INFO:LoggedFSStore:__setitem__('lon/.zarray', 313 BYTES) - 152.40456 ms
INFO:LoggedFSStore:__getitem__('lon/.zarray') -> 313 BYTES - 33.64609 ms
INFO:LoggedFSStore:setitems({'lon/0': b'\x02\x011\x04\xd4\x0) - 60.73121 ms
INFO:LoggedFSStore:__setitem__('lon/.zattrs', 160 BYTES) - 135.19151 ms
INFO:LoggedFSStore:__contains__('.zarray') - 64.26887 ms
INFO:LoggedFSStore:__contains__('.zgroup') -> True - 44.14023 ms
INFO:LoggedFSStore:__contains__('time/.zarray') - 65.71949 ms
INFO:LoggedFSStore:__contains__('time/.zgroup') - 73.62353 ms
INFO:LoggedFSStore:__setitem__('time/.zarray', 319 BYTES) - 130.98015 ms
INFO:LoggedFSStore:__getitem__('time/.zarray') -> 319 BYTES - 35.68954 ms
INFO:LoggedFSStore:setitems({'time/0': b'\x02\x01!\x08@[\x00) - 93.36168 ms
INFO:LoggedFSStore:__setitem__('time/.zattrs', 105 BYTES) - 128.76703 ms
CPU times: user 434 ms, sys: 54.4 ms, total: 489 ms
Wall time: 3.59 s
```

This shows that Xarray is adding a few calls of its own, but mostly the problem exists at the Zarr level.

**Comment by rabernat** [comment]
What does this all mean. It boils down to this

```
CPU times: user 571 ms, sys: 85.3 ms, total: 656 ms
Wall time: 5.31 s
```

The difference btw the 656ms of CPU time and the 5.31s of Wall time is time spent blocking on I/O. If we put all of these sequential I/O calls into an async event loop, I'm pretty sure we could eliminate most of that difference.

There is also absolutely no reason we need to be checking of `.zarray` and `.zgroup` over and over again...in series! 🤦  Nor do we need to be getting the data we have just set.

**Comment by guillaumeeb** [comment]
> @tinaok & @guillaumeeb - this is how Pangeo has made progress over the past years: carefully identifying performance bottlenecks and working systematically to fix them upstream in the open source libraries. If you have the time and interest, I really encourage you to dig into the zarr source code and see if you can optimize this a bit. 

Thanks so much @rabernat for all that.

Just wanted to say that I hope to do that soon, however I won't be available next week.

**Comment by martindurant** [comment]
I am thinking something like 
```
diff --git a/zarr/storage.py b/zarr/storage.py
index f5459990..0760d8d4 100644
--- a/zarr/storage.py
+++ b/zarr/storage.py
@@ -2858,7 +2858,11 @@ class ConsolidatedMetadataStore(Store):
         self.store = Store._ensure_store(store)

         # retrieve consolidated metadata
-        meta = json_loads(self.store[metadata_key])
+        try:
+            meta = json_loads(self.store[metadata_key])
+        except KeyError:
+            # new store for writing
+            meta = {'zarr_consolidated_format': 1}

         # check format of consolidated metadata
         consolidated_format = meta.get('zarr_consolidated_format', None)
@@ -2885,7 +2889,12 @@ class ConsolidatedMetadataStore(Store):
         raise ReadOnlyError()

     def __setitem__(self, key, value):
-        raise ReadOnlyError()
+        self.meta_store[key] = value
+
+    def commit(self):
+        out = {".zmetadata": self.meta_store}
+        out.update(self.meta_store)
+        del out["zarr_consolidated_format"]
+        self.store.setitems(out)

```
so that instead of calling consolidate_metadata, you use this store instead.

**Comment by rabernat** [comment]
That's one option. However, I see it as a bit of a workaround for the fundamental problem that the store has to be created in this very sequential, procedural way.

What if Zarr had a fast path to create an entire hierarchy at once, i.e.

```python
zarr.create_hierarchy(big_dict, store=store, mode="w")
```

and all the data were set using a single call to `store.setitems(...)`?

I feel like that would be a more elegant solution that would not require consolidated metadata. After all, the `aws s3' CLI test shows that it is possible to get all of this data in S3 (without consolidated metadata) in < 2s. Why shouldn't Zarr be able to do the same?

**Comment by martindurant** [comment]
Certainly. I don't have an immediate evaluation as to the merits of adding top level functions to zarr versus making more versatile storage layers. They end up doing the same thing here. What does your big_dict contain? 

setitems is (for now) only making any difference on FSStore compared to serial.

>  this very sequential, procedural way.

Some of this was also discussed in the async-zarr idea. That was about data access, possibly across multiple arrays (which zarr also can't normally do).

Note that fsspec instances also allow for "transactions" to defer finalising upload batches and make them semi-atomic. That wouldn't help with speed, but would with consistency, which any lazy/deferred collection of thing to write would need to worry about.

**Comment by martindurant** [comment]
(ah, and my diff doesn't actually require consolidated metadata, I am merely hijacking the class, because it was already doing some of this work; any dict could be used as the metadata store for uploading in a batch)
